src/lib/OccAnalysis.hs

```diff
@@ -0,0 +1,559 @@
+
+-- Run the given action with its own FV state, and return the FVs it
+-- accumulates for post-processing.
+isolated :: OCCM n a -> OCCM n (a, FV n)
+isolated action = do
+  r <- ask
+  lift11 $ lift11 $ runStateT1 (runReaderT1 r action) mempty
+
+-- Run the given action with its own FV state, and merge it afterward.
+-- This way the action can munge its state itself based on local
+-- information and know it's not affecting the surrounding state.
+locally :: OCCM n a -> OCCM n a
+locally action = do
+  (a, fvs) <- isolated action
+  modify (<> fvs)
+  return a
+
+-- Extend the IxExpr environment
+extend :: (BindsOneName b 'AtomNameC)
+  => b any n -> IxExpr n -> OCCM n a -> OCCM n a
+extend b ix = local (<> singletonNameMapE (binderName b) ix)
+
+-- Look up the `IxExpr` for a given name.  If the name doesn't occur in the map,
+-- we return a `Occ.Var` of the name itself, thus claiming that the name is its
+-- own summary.  This is what we want for `for` binders and top-level names.
+ixExpr :: AtomName n -> OCCM n (IxExpr n)
+ixExpr name = do
+  ixExprs <- ask
+  case lookupNameMapE name ixExprs of
+    Just ans -> return $ ans
+    Nothing -> return $ Occ.Var name
+
+-- `TabLamExpr` and `IxDicts` are meant to be inlined themselves, so we
+-- have to assume that their occurrences may be replicated many times,
+-- including statically.
+inlinedLater :: (HoistableE e) => e n -> OCCM n (e n)
+inlinedLater obj = do
+  (_, fvs) <- isolated $ countFreeVarsAsOccurrences obj
+  modify (<> useManyTimesStatic (useManyTimes fvs))
+  return obj
+
+-- === Computing IxExpr summaries ===
+
+summaryExpr :: Expr n -> OCCM n (IxExpr n)
+summaryExpr = \case
+  Atom atom -> summary atom
+  expr -> unknown expr
+
+summary :: Atom n -> OCCM n (IxExpr n)
+summary atom = case atom of
+  Var name -> ixExpr name
+  Lam _ -> invalid "Lam"
+  Con c -> constructor c
+  ACase _ _ _ -> error "Unexpected ACase outside of Simplify"
+  _ -> unknown atom
+  where
+    invalid tag = error $ "Unexpected indexing by " ++ tag
+    constructor = \case
+      -- TODO Represent the actual literal value?
+      Lit _ -> return $ Deterministic []
+      ProdCon elts -> Product <$> mapM summary elts
+      SumCon _ tag payload -> Inject tag <$> summary payload
+      SumAsProd _ (TagRepVal tag) payloads ->
+        Inject (fromIntegral tag) <$> (summary $ payloads !! (fromIntegral tag))
+      SumAsProd _ _ _ -> unknown atom
+      LabelCon _ -> invalid "LabelCon"
+      Newtype _ e -> summary e
+      BaseTypeRef _ -> invalid "BaseTypeRef"
+      TabRef _ -> invalid "TabRef"
+      ConRef _ -> invalid "ConRef"
+      ExplicitDict _ _ -> invalid "ExplicitDict"
+      DictHole _ _ -> invalid "DictHole"
+
+unknown :: HoistableE e => e n -> OCCM n (IxExpr n)
+unknown obj = do
+  freeIxs <- mapM ixExpr $ freeAtomVarsList obj
+  let freeIterationNames = nub $ concatMap freeAtomVarsList freeIxs
+  return $ Deterministic $ freeIterationNames
```

Comment:
For RWS I think so, because the reference will appear in the freeAtomVarsList if it is relevant, and I think that means the right thing will happen.  If the effect is IO, then I guess not.

However, this does mean that I have to detect wither and of the `freeIxs` are already `IxAll` and promote to `IxAll` if so.  Fixed that.